### ISSUE TITLE
Some issues for duplicating the paper results

### ISSUE THREAD

**Issue by wangcongcong123** [yumeng5/LOTClass#10]
Hi,

Thanks for open sourcing this.

When trying to duplicate the results as described in the paper, I encountered some issues. Much appreciated for some advice.

Taking the AGNews dataset as an example,

1. **About MCP Data Generation.** The default code shows if there are 20/50 overlapping indicative words, then there will be pseudo-labeled documents for MCP training. However, when first running `category_vocabulary` method, there will be less than 50 words (as specified by `args.category_vocab_size`) for each label since the `sorted_dict = {k:v for k, v in sorted(cat_dict.items(), key=lambda item: item[1], reverse=True)[:category_vocab_size]}` in `filter_keywords` (see [this line](https://github.com/yumeng5/LOTClass/blob/aa65ae2d249605adc28ad9650fdb867447b48375/src/trainer.py#L235)) is run at the beginning before the words filtering starts. Due to this, I changed `valid_idx = (match_count > match_threshold) & (input_mask > 0)` in [this line](https://github.com/yumeng5/LOTClass/blob/aa65ae2d249605adc28ad9650fdb867447b48375/src/trainer.py#L338) to `valid_idx = (match_count > len(category_vocab) * match_threshold / top_pred_num) & (input_mask > 0)`. Without this change, I got 0 pseudo-labeled documents for MCP. Now I can get thousands of pseudo-labeled documents for MCP. Hence, with this change, I move on to the next step in the pipeline - **Start MCP training**. Did I miss something if it works on your side without this change?

2. **About MCP Training.** With the pseudo-labeled documents generated from last step, I successfully started the MCP training. By looking at the loss as follows, it seemed nothing wrong.

```
Training model via masked category prediction.
Epoch 1:
100%|██████████| 873/873 [01:42<00:00,  8.51it/s]
Average training loss: 0.9677731394767761
  0%|          | 0/873 [00:00<?, ?it/s]Epoch 2:
100%|██████████| 873/873 [01:42<00:00,  8.49it/s]
Average training loss: 0.4305942952632904
  0%|          | 0/873 [00:00<?, ?it/s]Epoch 3:
100%|██████████| 873/873 [01:42<00:00,  8.49it/s]
Average training loss: 0.31928300857543945
```

However, some issues popped up when I was attempting to run the MCP-trained model at this stage for inference on the test set. The `inference` method always returns the `preds` with the same probability distribution across different examples (different input ids). Something like this: 

```
input_ids, input_mask, preds = self.inference(model, dataset_loader, rank, return_type="data")
print(preds)

tensor([[0.1776, 0.1679, 0.3113, 0.3432],
        [0.1776, 0.1679, 0.3113, 0.3432],
        [0.1776, 0.1679, 0.3113, 0.3432],
        ...,
        [0.1776, 0.1679, 0.3113, 0.3432],
        [0.1776, 0.1679, 0.3113, 0.3432],
        [0.1776, 0.1679, 0.3113, 0.3432]], device='cuda:0')
```
As a result, the MCP-trained model always gives random-guess mean accuracy (0.25 in the AGNews case). I am wondering if you guys have the same issue or I miss something? Much appreciation for any pointers.

3. **About Self Train.** Endeavoring to fix the issues as described in step 2, I hacked the code. The `preds` are the softmax probs over the logits of BERT's `[CLS]` last hidden state, so I am wondering if the `self_train` will solve the issue as it takes the `[CLS]` token into account. Unfortunately, the training still failed and the training log is as follows.

```
100%|██████████| 1600/1600 [03:04<00:00,  8.68it/s]
lr: 5.336e-07
Average training loss: 0.05783357098698616
Test acc: 0.25
100%|██████████| 1600/1600 [03:05<00:00,  8.63it/s]
lr: 9.925e-07
Average training loss: 0.016020258888602257
Test acc: 0.25
100%|██████████| 1600/1600 [03:05<00:00,  8.61it/s]
lr: 9.332e-07
Average training loss: 0.0071106418035924435
Test acc: 0.25
```
I have been stuck in these issues after many hours hacking for the code and ended up with no solutions. It would be really helpful for my research if there are some pointers or guides to solve these issues. Looking forward to reproduce the 0.864 test accuracy in AGNews as reported in the paper. Thanks.

FYI, my training environment is:

- Since I only have a budget-limited GPU on my desktop. I run the experiments on a single 6GB 2060 GPU. To fit into the memory, I set the batch size to be 8 with accumulation steps of 16. I kept the rest of hyper-parameters the same as in `agnews.sh`.

**Comment by yumeng5** [comment]
Hi @wangcongcong123,

Thanks for trying out the code and I'm sorry for the delay in my response--I've been very busy with several recent deadlines.

Actually, you don't need to change any code to run on the given four datasets. Specifically, you could simply run [`agnew.sh`](https://github.com/yumeng5/LOTClass/blob/master/agnews.sh) which has the correct hyperparameters set for you (you only need to change `train_batch_size=8`, `accum_steps=16`, `gpus=1` for your case). 

You could expect to see similar outputs in the terminal like the following for the AGNews dataset (your results might be slightly different due to different tie-breaking when multiple terms have the same frequency to replace the original category name):

```
Class 0 category vocabulary: ['politics', 'political', 'politicians', 'government', 'elections', 'politician', 'democracy',
 'democratic', 'governing', 'party', 'state', 'leadership', 'election', 'politically', 'affairs', 'issues', 'governments',
'voters', 'debate', 'cabinet', 'congress', 'democrat', 'administration', 'president', 'religion', 'republican', 'crisis', '
war', 'history', 'legislature', 'candidates', 'governance', 'pr', 'opposition', 'problems', 'relations', 'finance', 'justic
e', 'struggle', 'rhetoric', 'right', 'convention', 'violence', 'fighting', 'votes', 'senate', 'matters', 'fight', 'us', 'tr
ouble', 'republicans', 'parliament', 'one', 'conflict', 'soil', 'law', 'voting', 'representation', 'parliamentary', 'house'
, 'reality', 'campaign', 'wars', 'contest', 'candidate', 'campaigns', 'transition', 'legislative', 'question']

Class 1 category vocabulary: ['sports', 'games', 'sporting', 'athletics', 'game', 'national', 'news', 'athletic', 'espn', '
soccer', 'stadium', 'basketball', 'arts', 'racing', 'baseball', 'tv', 'hockey', 'pro', 'press', 'team', 'red', 'home', 'bay
', 'kings', 'legends', 'city', 'winning', 'miracle', 'olympic', 'go', 'giants', 'champions', 'ball', 'players', 'prime', 'b
oxing', 'teams', 'athletes', 'tennis', 'club', 'blue', 'coaches', 'gold', 'west', 'toronto', 'classic', 'pittsburgh', 'supe
r', 'nfl', 'magic', 'key', 'times', 'field', 'warriors', 'rogers', 'stars', 'gym', 'championship', 'losses', 'college', 'ml
b', 'veterans', 'rugby', 'hits', 'bc', 'sun', 'events', 'south', 'nba']

Class 2 category vocabulary: ['business', 'businesses', 'trade', 'commercial', 'enterprise', 'shop', 'money', 'market', 'co
mmerce', 'corporate', 'global', 'future', 'sales', 'general', 'group', 'retail', 'companies', 'management', 'operations', '
operation', 'corporation', 'store', 'division', 'firm', 'venture', 'brand', 'contract', 'revenue', 'economic', 'branch', 's
ubsidiary', 'personal', 'cash', 'short', 'line', 'bank', 'customer', 'concern', 'growth', 'chain', 'strategic', 'family', '
work', 'products', 'big', 'scientific', 'virtual', 'engineering', 'sector', 'trading', 'portfolio', 'ceo', 'segment', 'inve
stment', 'working', 'executive', 'private', 'services', 'public', 'job', 'marketing']

Class 3 category vocabulary: ['technology', 'technologies', 'tech', 'software', 'technological', 'device', 'equipment', 'ha
rdware', 'infrastructure', 'devices', 'system', 'knowledge', 'technique', 'digital', 'technical', 'concept', 'systems', 'ge
ar', 'techniques', 'material', 'functionality', 'process', 'facility', 'feature', 'capability', 'content', 'method', 'secur
ity', 'ability', 'network', 'internet', 'computing', 'chip', 'smart', 'modern', 'communication', 'language', 'mechanism', '
computer', 'design', 'cyber', 'standard', 'tool', 'development', 'format', 'protocol', 'wireless', 'phone', 'information',
'program', 'ce', 'plant', 'large', 'data', 'project', 'application', 'theory', 'science', 'performance', 'common', 'os', 'i
ct', 'speed', 'sensor', 'capabilities', 'electronic', 'society', 'silicon', 'invention', 'memory']
```

As shown above, each category has more than 50 representative words because the default [`category_vocab_size` is 100](https://github.com/yumeng5/LOTClass/blob/aa65ae2d249605adc28ad9650fdb867447b48375/src/train.py#L30).

Then you should see the MCP step locates category-indicative terms in the corpus and starts training, like the following:

```
Preparing self supervision for masked category prediction.
100%|████████████████████████████████████████████████████████████████████████████████████| 469/469 [05:38<00:00,  1.38it/s]
Number of documents with category indicative terms found for each category is: {0: 207, 1: 1365, 2: 1426, 3: 2049}
There are totally 5047 documents with category indicative terms.

Training model via masked category prediction.
Epoch 1:
100%|██████████████████████████████████████████████████████████████████████████████████████| 79/79 [00:35<00:00,  2.25it/s]
Average training loss: 0.8516192436218262
Epoch 2:
100%|██████████████████████████████████████████████████████████████████████████████████████| 79/79 [00:34<00:00,  2.26it/s]
Average training loss: 0.3009144365787506
Epoch 3:
100%|██████████████████████████████████████████████████████████████████████████████████████| 79/79 [00:34<00:00,  2.26it/s]
Average training loss: 0.15698106586933136
```

Note you should obtain a similar amount (might not be exactly the same due to the difference in the previous category vocabulary construction step) of category indicative terms found for each category like shown above. For example, the results obtained by [another user](https://github.com/yumeng5/LOTClass/issues/2#issuecomment-735352411) seems reasonable.

Please let me know if you are still having issues reproducing the results.

Best,
Yu

**Comment by wangcongcong123** [comment]
Hi Yu,

Thanks for your reply. It seems the loss looks fine on my side for MCP as well. I am wondering if you tried the self training step and what the accuracy on test set is like during the self training? Would you mind posting your self training log here? Cheers.

Regards,
Congcong

**Comment by yumeng5** [comment]
Hi Congcong,

I attached my self training log below. There will be some randomness in the final test set performance (like the following one gives me a 0.862 final test accuracy), but generally, you could expect to consistently get accuracies between 0.855-0.870 on the AGNews dataset.

```
Start self-training.
100%|████████████████████████████████████████████████████████████████████████████████████| 200/200 [02:29<00:00,  1.33it/s]
lr: 5.336e-07
Average training loss: 0.1297808289527893
Test acc: 0.8336842060089111
100%|████████████████████████████████████████████████████████████████████████████████████| 200/200 [02:33<00:00,  1.30it/s]
lr: 9.925e-07
Average training loss: 0.09482331573963165
Test acc: 0.8452631831169128
100%|████████████████████████████████████████████████████████████████████████████████████| 200/200 [02:29<00:00,  1.34it/s]
lr: 9.332e-07
Average training loss: 0.09001187980175018
Test acc: 0.8526315689086914
100%|████████████████████████████████████████████████████████████████████████████████████| 200/200 [02:34<00:00,  1.29it/s]
lr: 8.739e-07
Average training loss: 0.07647929340600967
Test acc: 0.8552631139755249
100%|████████████████████████████████████████████████████████████████████████████████████| 200/200 [02:33<00:00,  1.30it/s]
lr: 8.147e-07
Average training loss: 0.0710068792104721
Test acc: 0.8584210276603699
100%|████████████████████████████████████████████████████████████████████████████████████| 200/200 [02:33<00:00,  1.31it/s]
lr: 7.554e-07
Average training loss: 0.06767375767230988
Test acc: 0.8602631688117981
100%|████████████████████████████████████████████████████████████████████████████████████| 200/200 [02:32<00:00,  1.31it/s]
lr: 6.961e-07
Average training loss: 0.06254400312900543
Test acc: 0.8610526323318481
100%|████████████████████████████████████████████████████████████████████████████████████| 200/200 [02:33<00:00,  1.31it/s]
lr: 6.368e-07
Average training loss: 0.05620194971561432
Test acc: 0.8622368574142456
100%|████████████████████████████████████████████████████████████████████████████████████| 200/200 [02:32<00:00,  1.31it/s]
lr: 5.775e-07
Average training loss: 0.0571054071187973
Test acc: 0.8609210848808289
100%|████████████████████████████████████████████████████████████████████████████████████| 200/200 [02:31<00:00,  1.32it/s]
lr: 5.182e-07
Average training loss: 0.05499818176031113
Test acc: 0.8636841773986816
100%|████████████████████████████████████████████████████████████████████████████████████| 200/200 [02:30<00:00,  1.33it/s]
lr: 4.589e-07
Average training loss: 0.05416369065642357
Test acc: 0.863289475440979
100%|████████████████████████████████████████████████████████████████████████████████████| 200/200 [02:28<00:00,  1.34it/s]
lr: 3.996e-07
Average training loss: 0.05627661198377609
Test acc: 0.8631579279899597
100%|████████████████████████████████████████████████████████████████████████████████████| 200/200 [02:27<00:00,  1.36it/s]
lr: 3.403e-07
Average training loss: 0.054078616201877594
Test acc: 0.8631578683853149
100%|████████████████████████████████████████████████████████████████████████████████████| 200/200 [02:30<00:00,  1.32it/s]
lr: 2.81e-07
Average training loss: 0.05103151500225067
Test acc: 0.8630263209342957
100%|████████████████████████████████████████████████████████████████████████████████████| 200/200 [02:28<00:00,  1.35it/s]
lr: 2.217e-07
Average training loss: 0.0507686622440815
Test acc: 0.8636841773986816
100%|████████████████████████████████████████████████████████████████████████████████████| 200/200 [02:29<00:00,  1.34it/s]
lr: 1.625e-07
Average training loss: 0.049498897045850754
Test acc: 0.8640789985656738
100%|████████████████████████████████████████████████████████████████████████████████████| 200/200 [02:29<00:00,  1.34it/s]
lr: 1.032e-07
Average training loss: 0.04837364703416824
Test acc: 0.8630262613296509
100%|████████████████████████████████████████████████████████████████████████████████████| 200/200 [02:29<00:00,  1.33it/s]
lr: 4.388e-08
Average training loss: 0.05135683715343475
Test acc: 0.8621052503585815
Saving final model to datasets/agnews/final_model.pt
```

Please let me know if you have any other questions/concerns.

Best,
Yu
